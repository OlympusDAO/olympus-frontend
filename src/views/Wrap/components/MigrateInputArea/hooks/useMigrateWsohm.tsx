import { t } from "@lingui/macro";
import { ContractReceipt } from "ethers";
import { useMutation, useQueryClient } from "react-query";
import { useDispatch } from "react-redux";
import { GOHM_ADDRESSES, MIGRATOR_ADDRESSES, WSOHM_ADDRESSES } from "src/constants/addresses";
import { DecimalBigNumber } from "src/helpers/DecimalBigNumber/DecimalBigNumber";
import { useWeb3Context } from "src/hooks";
import { balanceQueryKey, useBalance } from "src/hooks/useBalance";
import { useDynamicMigratorContract } from "src/hooks/useContract";
import { useTestableNetworks } from "src/hooks/useTestableNetworks";
import { error as createErrorToast, info as createInfoToast } from "src/slices/MessagesSlice";

export const useMigrateWsohm = () => {
  const dispatch = useDispatch();
  const client = useQueryClient();
  const networks = useTestableNetworks();
  const balances = useBalance(WSOHM_ADDRESSES);
  const { address, networkId } = useWeb3Context();
  const contract = useDynamicMigratorContract(MIGRATOR_ADDRESSES, true);

  return useMutation<ContractReceipt, Error, string>(
    async amount => {
      if (!amount || isNaN(Number(amount))) throw new Error(t`Please enter a number`);

      const _amount = new DecimalBigNumber(amount, 18);

      if (!_amount.gt(new DecimalBigNumber("0", 18))) throw new Error(t`Please enter a number greater than 0`);

      if (!contract || (networkId !== networks.AVALANCHE && networkId !== networks.ARBITRUM))
        throw new Error(t`Please switch to the Abritrum or Avalanche networks to migrate`);

      const balance = balances[networkId].data;

      if (!balance) throw new Error(t`Please refresh your page and try again`);

      if (_amount.gt(balance)) throw new Error(t`You cannot migrate more than your wsOHM balance`);

      const transaction = await contract.migrate(_amount.toBigNumber());
      return transaction.wait();
    },
    {
      onError: error => {
        dispatch(createErrorToast(error.message));
      },
      onSuccess: async () => {
        const keysToRefetch = [
          balanceQueryKey(address, WSOHM_ADDRESSES, networkId),
          balanceQueryKey(address, GOHM_ADDRESSES, networkId),
        ];

        const promises = keysToRefetch.map(key => client.refetchQueries(key, { active: true }));

        await Promise.all(promises);

        dispatch(createInfoToast(t`Successfully migrated from wsOHM to gOHM`));
      },
    },
  );
};
