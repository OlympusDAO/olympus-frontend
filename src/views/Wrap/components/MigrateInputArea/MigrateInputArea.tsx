import { t, Trans } from "@lingui/macro";
import { Box, FormControl, Grid, MenuItem, Select, Typography } from "@material-ui/core";
import { Input, PrimaryButton } from "@olympusdao/component-library";
import { useState } from "react";
import { TokenAllowanceGuard } from "src/components/TokenAllowanceGuard/TokenAllowanceGuard";
import { MIGRATOR_ADDRESSES, WSOHM_ADDRESSES } from "src/constants/addresses";
import { assert } from "src/helpers/types/assert";
import { useWeb3Context } from "src/hooks";
import { useBalance } from "src/hooks/useBalance";
import { useTestableNetworks } from "src/hooks/useTestableNetworks";

import { useMigrateWsohm } from "./hooks/useMigrateWsohm";

export const MigrateInputArea = () => {
  const networks = useTestableNetworks();
  const { networkId } = useWeb3Context();

  // Max balance stuff
  assert(
    networkId === networks.ARBITRUM || networkId === networks.AVALANCHE,
    "Component should only be mounted when connected to Arbitrum or Avalanche",
  );
  const [amount, setAmount] = useState("");
  const balance = useBalance(WSOHM_ADDRESSES)[networkId].data;
  const setMax = () => balance && setAmount(balance.toString());

  // Mutation stuff
  const migrateMutation = useMigrateWsohm();
  const handleSubmit = (event: React.FormEvent<WrapFormElement>) => {
    event.preventDefault();
    migrateMutation.mutate(amount);
  };

  return (
    <Box mt={2} mb={4}>
      <Box display="flex" alignItems="center">
        <Typography>Migrate</Typography>

        <FormControl style={{ margin: "0 10px" }}>
          <Select label="Asset" disableUnderline id="asset-select" value="wsOHM">
            <MenuItem value="wsOHM">wsOHM</MenuItem>
          </Select>
        </FormControl>

        <Typography>
          <span className="asset-select-label"> to </span>
        </Typography>

        <FormControl style={{ margin: "0 10px" }}>
          <Select value="gOHM" label="Asset" disableUnderline id="asset-select">
            <MenuItem value="gOHM">gOHM</MenuItem>
          </Select>
        </FormControl>
      </Box>

      <Box my={1}>
        <TokenAllowanceGuard
          tokenAddressMap={WSOHM_ADDRESSES}
          spenderAddressMap={MIGRATOR_ADDRESSES}
          message={
            <>
              <Trans>Please approve Olympus DAO to use your</Trans> <b>wsOHM</b> <Trans>for migrating</Trans>.
            </>
          }
        >
          <form onSubmit={handleSubmit} className="stake-tab-panel wrap-page">
            <Grid container>
              <Grid item xs={12} sm={8} style={{ paddingRight: "4px" }}>
                <Input
                  labelWidth={0}
                  value={amount}
                  id="amount-input"
                  endString={t`Max`}
                  name="amount-input"
                  endStringOnClick={setMax}
                  placeholder={t`Enter an amount of wsOHM`}
                  disabled={migrateMutation.isLoading}
                  onChange={event => setAmount(event.target.value)}
                />
              </Grid>

              <Grid item xs={12} sm={4}>
                <Box mt={[1, 0]}>
                  <PrimaryButton
                    fullWidth
                    type="submit"
                    className=""
                    disabled={migrateMutation.isLoading}
                    data-testid="migrate-button"
                  >
                    {migrateMutation.isLoading ? "Migrating..." : "Migrate"}
                  </PrimaryButton>
                </Box>
              </Grid>
            </Grid>
          </form>
        </TokenAllowanceGuard>
      </Box>
    </Box>
  );
};

interface WrapFormElement extends HTMLFormElement {
  elements: HTMLFormControlsCollection & { "amount-input": HTMLInputElement };
}
