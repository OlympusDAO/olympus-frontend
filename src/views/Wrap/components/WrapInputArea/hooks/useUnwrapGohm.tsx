import { t } from "@lingui/macro";
import { ContractReceipt } from "ethers";
import { useMutation, useQueryClient } from "react-query";
import { useDispatch } from "react-redux";
import { GOHM_ADDRESSES, SOHM_ADDRESSES, STAKING_ADDRESSES } from "src/constants/addresses";
import { DecimalBigNumber } from "src/helpers/DecimalBigNumber/DecimalBigNumber";
import { useWeb3Context } from "src/hooks";
import { balanceQueryKey, useBalance } from "src/hooks/useBalance";
import { useDynamicStakingContract } from "src/hooks/useContract";
import { useTestableNetworks } from "src/hooks/useTestableNetworks";
import { error as createErrorToast, info as createInfoToast } from "src/slices/MessagesSlice";

export const useUnwrapGohm = () => {
  const dispatch = useDispatch();
  const client = useQueryClient();
  const { address } = useWeb3Context();
  const networks = useTestableNetworks();
  const balance = useBalance(GOHM_ADDRESSES)[networks.MAINNET].data;
  const contract = useDynamicStakingContract(STAKING_ADDRESSES, true);

  return useMutation<ContractReceipt, Error, string>(
    async amount => {
      if (!amount || isNaN(Number(amount))) throw new Error(t`Please enter a number`);

      const _amount = new DecimalBigNumber(amount, 18);

      if (!_amount.gt(new DecimalBigNumber("0", 18))) throw new Error(t`Please enter a number greater than 0`);

      if (!balance) throw new Error(t`Please refresh your page and try again`);

      if (_amount.gt(balance)) throw new Error(t`You cannot unwrap more than your gOHM balance`);

      if (!contract) throw new Error(t`Please switch to the Ethereum network to unwrap your gOHM`);

      if (!address) throw new Error(t`Please refresh your page and try again`);

      const transaction = await contract.unwrap(address, _amount.toBigNumber());
      return transaction.wait();
    },
    {
      onError: error => {
        dispatch(createErrorToast(error.message));
      },
      onSuccess: async () => {
        const keysToRefetch = [
          balanceQueryKey(address, SOHM_ADDRESSES, networks.MAINNET),
          balanceQueryKey(address, GOHM_ADDRESSES, networks.MAINNET),
        ];

        const promises = keysToRefetch.map(key => client.refetchQueries(key, { active: true }));

        await Promise.all(promises);

        dispatch(createInfoToast(t`Successfully unwrapped from gOHM to sOHM`));
      },
    },
  );
};
