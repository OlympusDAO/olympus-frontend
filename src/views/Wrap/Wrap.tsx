import { t } from "@lingui/macro";
import { Box, Divider, Grid, Link, Typography } from "@mui/material";
import { Icon, MetricCollection, Paper } from "@olympusdao/component-library";
import { WalletConnectedGuard } from "src/components/WalletConnectedGuard";
import { useWeb3Context } from "src/hooks";
import { useTestableNetworks } from "src/hooks/useTestableNetworks";

import { CurrentIndex, GOHMPrice, SOHMPrice } from "../TreasuryDashboard/components/Metric/Metric";
import { MigrateInputArea } from "./components/MigrateInputArea/MigrateInputArea";
import { WrapBalances } from "./components/WrapBalances";
import { WrapInputArea } from "./components/WrapInputArea/WrapInputArea";
import { WrapSwitchNetwork } from "./components/WrapSwitchNetwork";

const Wrap: React.FC = () => {
  const networks = useTestableNetworks();
  const { networkId } = useWeb3Context();
  const isMigrating = networkId === networks.ARBITRUM || networkId === networks.AVALANCHE;

  return (
    <div id="stake-view">
      <Paper headerText={t`Wrap / Unwrap`} topRight={<GOHMExternalLink />}>
        <Box mb="28px">
          <Grid>
            <MetricCollection>
              <SOHMPrice />
              <CurrentIndex />
              <GOHMPrice />
            </MetricCollection>
          </Grid>
        </Box>

        <WalletConnectedGuard message="Connect your wallet to wrap/unwrap your staked tokens">
          {isMigrating ? <MigrateInputArea /> : <WrapInputArea />}

          <WrapBalances />

          <Divider />

          <Box width="100%" p={1} sx={{ textAlign: "center" }}>
            <WrapSwitchNetwork />
          </Box>
        </WalletConnectedGuard>
      </Paper>
    </div>
  );
};

const GOHMExternalLink = () => (
  <Link
    target="_blank"
    aria-label="wsohm-wut"
    style={{ textDecoration: "none" }}
    href="https://docs.olympusdao.finance/main/contracts/tokens#gohm"
  >
    <Box display="flex" alignItems="center">
      <Typography>gOHM</Typography> <Icon style={{ marginLeft: "5px" }} name="arrow-up" />
    </Box>
  </Link>
);

export default Wrap;
