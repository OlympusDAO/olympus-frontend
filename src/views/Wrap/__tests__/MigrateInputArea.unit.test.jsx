import { fireEvent } from "@testing-library/react";
import { BigNumber } from "ethers";
import Messages from "src/components/Messages/Messages";
import { useContractAllowance } from "src/hooks/useContractAllowance";
import * as useWeb3Context from "src/hooks/web3Context";
import { mockWeb3Context } from "src/testHelpers";

import { act, render, screen } from "../../../testUtils";
import { MigrateInputArea } from "../components/MigrateInputArea/MigrateInputArea";
import Wrap from "../Wrap";

jest.mock("src/hooks/useContractAllowance");
let container;

beforeEach(async () => {
  const data = jest.spyOn(useWeb3Context, "useWeb3Context");
  useContractAllowance.mockReturnValue({ data: BigNumber.from(10000) });
  data.mockReturnValue({
    ...mockWeb3Context,
    networkId: 43114,
  });

  await act(async () => {
    ({ container } = render(
      <>
        <Messages />
        <Wrap />
      </>,
    ));
  });
});

afterEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe("Wrap Input Area", () => {
  it("Should Render Input when has Token Approval", async () => {
    fireEvent.change(await screen.findByPlaceholderText("Enter an amount of wsOHM"), { target: { value: "1" } });
    expect(screen.getByTestId("migrate-button"));
    expect(container).toMatchSnapshot();
  });
});

describe("Check Migrate to gOHM Error Messages", () => {
  it("Should error when not on Avalanche or Arbitrum", async () => {
    const data = jest.spyOn(useWeb3Context, "useWeb3Context");
    data.mockReturnValue({
      ...mockWeb3Context,
      networkId: 137, //polygon isnt supported for wsOHM to gOHM
    });
    expect(() => render(<MigrateInputArea />)).toThrowError(
      "Component should only be mounted when connected to Arbitrum or Avalanche",
    );
  });
  it("Error message with no amount", async () => {
    fireEvent.click(screen.getByTestId("migrate-button"));
    expect(await screen.findByText("Please enter a number")).toBeInTheDocument();
  });

  it("Error message with amount <=0", async () => {
    fireEvent.change(await screen.findByPlaceholderText("Enter an amount of wsOHM"), { target: { value: "-1" } });
    fireEvent.click(screen.getByTestId("migrate-button"));
    expect(await screen.findByText("Please enter a number greater than 0")).toBeInTheDocument();
  });

  // it("Error message amount > 0 but no wallet balance", async () => {
  //   fireEvent.change(await screen.findByPlaceholderText("Enter an amount of wsOHM"), { target: { value: "10000" } });
  //   fireEvent.click(screen.getByTestId("migrate-button"));
  //   expect(
  //     (await screen.findByText("Please refresh your page and try again")) ||
  //       (await screen.findByText("You cannot migrate more than your wsOHM balance")),
  //   ).toBeInTheDocument();
  // });
});
