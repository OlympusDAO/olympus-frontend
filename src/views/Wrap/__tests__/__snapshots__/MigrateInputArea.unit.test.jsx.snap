// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrap Input Area Should Render Input when has Token Approval 1`] = `
<div>
  <div>
    <div />
  </div>
  <div
    id="stake-view"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 Paper-root  css-m19e0g-MuiPaper-root"
      style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1y6dg9a-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item card-header css-159wrwf-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-gg4vpm"
          >
            <div
              class="MuiBox-root css-1xhj18k"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 header-text css-903ewo-MuiTypography-root"
              >
                Wrap / Unwrap
              </h5>
            </div>
            <div
              class="top-right"
            >
              <a
                aria-label="wsohm-wut"
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-ao3601-MuiTypography-root-MuiLink-root"
                href="https://docs.olympusdao.finance/main/contracts/tokens#gohm"
                style="text-decoration: none;"
                target="_blank"
              >
                <div
                  class="MuiBox-root css-70qvj9"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-69sme-MuiTypography-root"
                  >
                    gOHM
                  </p>
                   
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1fx2bp7-MuiSvgIcon-root"
                    focusable="false"
                    style="margin-left: 5px;"
                    viewBox="0 0 20 20"
                  >
                    <path
                      d="M4.297 17.445h9.539c1.523 0 2.305-.78 2.305-2.28v-9.58c0-1.507-.782-2.288-2.305-2.288h-9.54c-1.523 0-2.304.773-2.304 2.289v9.578c0 1.508.781 2.281 2.305 2.281Zm.016-.968c-.875 0-1.352-.461-1.352-1.368V5.633c0-.899.477-1.367 1.352-1.367h9.5c.867 0 1.359.468 1.359 1.367v9.476c0 .907-.492 1.368-1.36 1.368h-9.5Zm7.296-4.235c.266 0 .438-.195.438-.476V7.867c0-.344-.188-.492-.492-.492H7.64c-.29 0-.47.172-.47.438 0 .265.188.445.477.445H9.53l1.133-.078-1.055.992-3.382 3.383a.476.476 0 0 0-.149.328c0 .273.18.453.453.453a.47.47 0 0 0 .344-.149L10.25 9.82l.984-1.047-.078 1.282v1.718c0 .29.18.47.453.47Z"
                    />
                  </svg>
                </div>
              </a>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-159wrwf-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-15i1vgz"
          >
            <div
              class="MuiGrid-root css-vz6rvy-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-rjww1q"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1wlo379-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wmkgw3-MuiGrid-root"
                  >
                    <div
                      class="Metric-root  css-1xkrsym"
                    >
                      <div
                        class="MuiBox-root css-g45inh"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-label css-tlko1g-MuiTypography-root"
                        >
                          sOHM Price
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-metric css-69sme-MuiTypography-root"
                          style="width: 100%;"
                        >
                          <span
                            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
                            style="width: 100%;"
                          />
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wmkgw3-MuiGrid-root"
                  >
                    <div
                      class="Metric-root  css-1xkrsym"
                    >
                      <div
                        class="MuiBox-root css-g45inh"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-label css-tlko1g-MuiTypography-root"
                        >
                          Current Index
                          <div
                            class="MuiBox-root css-w341jy"
                            style="font-size: 14px;"
                          >
                            <div
                              class="MuiBox-root css-0"
                              style="display: inline-flex; justify-content: center; align-self: center;"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall info-icon css-1fx2bp7-MuiSvgIcon-root"
                                focusable="false"
                                style="margin: 0px 5px; font-size: 1em;"
                                viewBox="0 0 20 20"
                              >
                                <path
                                  d="M 10 20 C 15.475 20 20 15.473 20 10 C 20 4.518 15.473 0 9.991 0 C 4.516 0 0 4.518 0 10 C 0 15.475 4.527 20 10 20 Z M 10 18.705 C 5.189 18.714 1.287 14.812 1.297 10.001 C 1.28 5.189 5.179 1.281 9.991 1.285 C 14.807 1.28 18.714 5.182 18.714 9.999 C 18.719 14.811 14.813 18.712 10 18.702 Z M 9.941 6.242 C 10.559 6.242 11.038 5.763 11.038 5.156 C 11.043 4.547 10.549 4.053 9.94 4.058 C 9.334 4.057 8.842 4.55 8.844 5.156 C 8.843 5.761 9.337 6.249 9.941 6.242 Z M 8.216 15.444 L 12.303 15.444 C 12.623 15.444 12.871 15.204 12.871 14.895 C 12.87 14.584 12.615 14.334 12.303 14.338 L 10.868 14.338 L 10.868 8.754 C 10.868 8.346 10.658 8.065 10.269 8.065 L 8.345 8.065 C 7.919 8.045 7.631 8.493 7.826 8.872 C 7.925 9.065 8.128 9.182 8.345 9.172 L 9.652 9.172 L 9.652 14.338 L 8.216 14.338 C 7.905 14.334 7.649 14.584 7.648 14.895 C 7.648 15.204 7.897 15.444 8.216 15.444 Z"
                                />
                              </svg>
                            </div>
                          </div>
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-metric css-69sme-MuiTypography-root"
                          style="width: 100%;"
                        >
                          <span
                            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
                            style="width: 100%;"
                          />
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wmkgw3-MuiGrid-root"
                  >
                    <div
                      class="Metric-root  css-1xkrsym"
                    >
                      <div
                        class="MuiBox-root css-g45inh"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-label css-tlko1g-MuiTypography-root"
                        >
                          gOHM Price
                          <div
                            class="MuiBox-root css-w341jy"
                            style="font-size: 14px;"
                          >
                            <div
                              class="MuiBox-root css-0"
                              style="display: inline-flex; justify-content: center; align-self: center;"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall info-icon css-1fx2bp7-MuiSvgIcon-root"
                                focusable="false"
                                style="margin: 0px 5px; font-size: 1em;"
                                viewBox="0 0 20 20"
                              >
                                <path
                                  d="M 10 20 C 15.475 20 20 15.473 20 10 C 20 4.518 15.473 0 9.991 0 C 4.516 0 0 4.518 0 10 C 0 15.475 4.527 20 10 20 Z M 10 18.705 C 5.189 18.714 1.287 14.812 1.297 10.001 C 1.28 5.189 5.179 1.281 9.991 1.285 C 14.807 1.28 18.714 5.182 18.714 9.999 C 18.719 14.811 14.813 18.712 10 18.702 Z M 9.941 6.242 C 10.559 6.242 11.038 5.763 11.038 5.156 C 11.043 4.547 10.549 4.053 9.94 4.058 C 9.334 4.057 8.842 4.55 8.844 5.156 C 8.843 5.761 9.337 6.249 9.941 6.242 Z M 8.216 15.444 L 12.303 15.444 C 12.623 15.444 12.871 15.204 12.871 14.895 C 12.87 14.584 12.615 14.334 12.303 14.338 L 10.868 14.338 L 10.868 8.754 C 10.868 8.346 10.658 8.065 10.269 8.065 L 8.345 8.065 C 7.919 8.045 7.631 8.493 7.826 8.872 C 7.925 9.065 8.128 9.182 8.345 9.172 L 9.652 9.172 L 9.652 14.338 L 8.216 14.338 C 7.905 14.334 7.649 14.584 7.648 14.895 C 7.648 15.204 7.897 15.444 8.216 15.444 Z"
                                />
                              </svg>
                            </div>
                          </div>
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1 Metric-metric css-69sme-MuiTypography-root"
                          style="width: 100%;"
                        >
                          <span
                            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
                            style="width: 100%;"
                          />
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-xwfmks"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-69sme-MuiTypography-root"
              >
                Migrate
              </p>
              <div
                class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
                style="margin: 0px 10px;"
              >
                <div
                  class="MuiInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1b50vqc-MuiInputBase-root-MuiInput-root-MuiSelect-root"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby="asset-select"
                    class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-wa1hpy-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                    id="asset-select"
                    role="button"
                    tabindex="0"
                  >
                    wsOHM
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    tabindex="-1"
                    value="wsOHM"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiSelect-icon MuiSelect-iconStandard css-1xyntl6-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 20 20"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body1 css-69sme-MuiTypography-root"
              >
                <span
                  class="asset-select-label"
                >
                   to 
                </span>
              </p>
              <div
                class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
                style="margin: 0px 10px;"
              >
                <div
                  class="MuiInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1b50vqc-MuiInputBase-root-MuiInput-root-MuiSelect-root"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby="asset-select"
                    class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-wa1hpy-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                    id="asset-select"
                    role="button"
                    tabindex="0"
                  >
                    gOHM
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    tabindex="-1"
                    value="gOHM"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiSelect-icon MuiSelect-iconStandard css-1xyntl6-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 20 20"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root css-1xlzx9v"
            >
              <div
                class="css-1jiamkk"
              >
                <form
                  class="stake-tab-panel wrap-page"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-1bxbb5o-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-111wfws-MuiGrid-root"
                      style="padding-right: 4px;"
                    >
                      <div
                        class="MuiFormControl-root MuiFormControl-fullWidth Input-root css-3tw635-MuiFormControl-root"
                      >
                        <div
                          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-eqe2qg-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1vvkud8-MuiInputBase-input-MuiOutlinedInput-input"
                            id="amount-input"
                            name="amount-input"
                            placeholder="Enter an amount of wsOHM"
                            type="text"
                            value="1"
                          />
                          <div
                            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                          >
                            <button
                              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-rmjmo8-MuiButtonBase-root-MuiButton-root"
                              tabindex="0"
                              type="button"
                            >
                              Max
                            </button>
                          </div>
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1csa3kb-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-hdw1oc"
                            >
                              <span
                                class="notranslate"
                              >
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wmkgw3-MuiGrid-root"
                    >
                      <div
                        class="MuiBox-root css-w4egcq"
                      >
                        <button
                          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root  css-8aw5i7-MuiButtonBase-root-MuiButton-root"
                          data-testid="migrate-button"
                          style="height: 43px;"
                          tabindex="0"
                          type="submit"
                        >
                          Migrate
                        </button>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
          <div
            class="DataRow-root data-row css-19fvmib"
          >
            <div
              class="MuiBox-root css-1xhj18k"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-4xvy19-MuiTypography-root"
              >
                wsOHM Balance (Avalanche)
              </p>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-4xvy19-MuiTypography-root"
              style="text-align: right;"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
                style="width: 80px;"
              />
            </p>
          </div>
          <div
            class="DataRow-root data-row css-19fvmib"
          >
            <div
              class="MuiBox-root css-1xhj18k"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-4xvy19-MuiTypography-root"
              >
                gOHM Balance (Avalanche)
              </p>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-4xvy19-MuiTypography-root"
              style="text-align: right;"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
                style="width: 80px;"
              />
            </p>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-na3zg8"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-rg0wkx-MuiTypography-root"
              style="margin: 15px 0px 10px 0px;"
            >
              Back to Ethereum Mainnet
            </h6>
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-1sigo3z-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="Ethereum Logo"
                height="28px"
                src="wETH.svg"
                width="28px"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6 css-rg0wkx-MuiTypography-root"
                style="margin-left: 8px;"
              >
                Ethereum
              </h6>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
